### PR TITLE
Add error span for unknown local call

### PR DESCRIPTION
**Pull request by viniciusmuller**
Adds an error span to unknown local calls, only up to the end of the function name, since I couldn't find a reliable way to track whether the function call has parens or not.

Currently it works fine but the test I've added is failing, when I execute it normally using the compiler, we have the column information reported:

```ex
# unknown-local-call.ex
defmodule Mod do
  def func do
    _result = my_unknown_func_call(10, 20)
  end
end
```

`make && elixir unknown-local-call.ex`
```ex
    error: undefined function my_unknown_func_call/2 (expected Mod to define such a function or for it to be imported, but none are available)
    │
  3 │     _result = my_unknown_func_call(10, 20)
    │               ^^^^^^^^^^^^^^^^^^^^
    │
    └─ unknown-local-call.ex:3:15: Mod.func/0

** (CompileError) unknown-local-call.ex: cannot compile module Mod (errors have been logged)
``` 

When using a file annotation, we lose the column information:

```ex
# test.ex
source = """
defmodule Mod do
  @file "test.ex"
  def func do
    _result = my_unknown_func_call(10, 20)
  end
end
"""

ast = Code.string_to_quoted!(source, columns: true)
Code.eval_quoted(ast)
```

`make && elixir test.ex`
```ex
    error: undefined function my_unknown_func_call/2 (expected Mod to define such a function or for it to be imported, but none are available)
    │
  4 │   def func do
    │   ^^^^^^^^^^^
    │
    └─ test.ex:4: Mod.func/0

** (CompileError) nofile: cannot compile module Mod (errors have been logged)
    nofile:1: (file)
```

Showing the wrong line here is to be expected, but notice that if we add the `\@file` annotation to the function when we execute it using `Code.eval_quoted` we lose the column information in the process, even though the ast generated by `Code.string_to_quoted` has it.

### CODE REVIEW

File: lib/elixir/lib/module/locals_tracker.ex
Comment (josevalim) [inline lib/elixir/lib/module/locals_tracker.ex:108]: Nitpick: unpack the local here:

```suggestion
          {{name, _arity} = local, position, macro_dispatch?} <- out_neighbours(bag, {:local, pair}),
```

**Comment by josevalim** [comment]
> When using a file annotation, we lose the column information:

Correct. Because we only use the `@file` annotation when the code is meta-programmed.  What would be more interesting is if, when using `EEx.function_from_file`, then we show the column span from the EEx template, that's what we should strive for!

**Review by josevalim — APPROVED** [review]
Beautifully done!

**Comment by viniciusmuller** [comment]
Could you elaborate on the `function_from_file`? I couldn't get it to report column information as well:

```ex
# file.ex
defmodule TestFunctionFromFile do
  require EEx

  # --> broken-template.eex
  # <%= _result = unknown_function_call() %>

  EEx.function_from_file(:def, :sample, "broken-template.eex", [], parser_options: [columns: true])
end
```

```
# make && elixir file.ex
 
    error: undefined function unknown_function_call/0 (expected TestFunctionFromFile to define such a function or for it to be imported, but none are available)
    │
  1 │ <%= _result = unknown_function_call() %>
    │ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    │
    └─ broken-template.eex:1: TestFunctionFromFile.sample/0

** (CompileError) go.ex: cannot compile module TestFunctionFromFile (errors have been logged)

```

File: lib/elixir/test/elixir/module/locals_tracker_test.exs
Comment (viniciusmuller) [inline lib/elixir/test/elixir/module/locals_tracker_test.exs:56]: I guess we can filter span out of here as well to keep just the position, but this should be inoffensive as well since it's just a meta

**Comment by josevalim** [comment]
> Could you elaborate on the function_from_file? I couldn't get it to report column information as well:

There may be deeper issues on the EEx one then, perhaps we are not forwarding the column correctly when we parse Elixir code within a template. We can debug it in the future. :)

File: lib/elixir/test/elixir/module/locals_tracker_test.exs
Comment (josevalim) [reply lib/elixir/test/elixir/module/locals_tracker_test.exs:56]: This is private API, so we don't care. Having it here is fine. :)

**Comment by josevalim** [comment]
To me this is ready to go, we can handle EEx in another PR. :)

**Comment by josevalim** [comment]
CI is failing. :)

**Comment by viniciusmuller** [comment]
Yeah, it was failing because we couldn't find a proper way to test this, since EEx also didn't work.
I'll push a commit that updates the test to pass and adds a TODO, but I think we can also wait until we get the test working before merging

I was re-reading and thinking
> Correct. Because we only use the @file annotation when the code is meta-programmed. What would be more interesting is if, when using EEx.function_from_file, then we show the column span from the EEx template, that's what we should strive for!

We are using `@file` successfully in all the other tests, this is the only case where we are actually losing the column information when we add the file annotation, and we also lose it on EEx.

But if we just run the compiler manually, we can see that we have the column information and the span works nicely

**Comment by josevalim** [comment]
Ok then, I know what it is then. We have some code in elixir_def that sees if `@file` was given and it assumes that the code was generated and that is likely removing the line+column information? We can try to fix that here or in a separate PR, your call. :)

**Comment by viniciusmuller** [comment]
> Ok then, I know what it is then. We have some code in elixir_def that sees if `@file` was given and it assumes that the code was generated and that is likely removing the line+column information? We can try to fix that here or in a separate PR, your call. :)

Ok, I can take a look into this, probably it will take a bit to get used to these internals but we can open another PR fixing it and then update this one

Also, when you talked about EEx you meant just having a test case that allows us to assert that we're reporting correctly in this case as well? If so, we can also add this here, just need to fix the file attribute issue first

**Comment by josevalim** [comment]
We can add it to EEx test suite. It is you if in this or another PR. :)

**Comment by josevalim** [comment]
Just checking: is this pending anything from me? :)

**Comment by viniciusmuller** [comment]
Not really @josevalim, I just need to find some time to continue working on it

If you could try to debug the current blocker of losing the column information when using eex templates and the functions from `Code` it would be a great help and I believe this could be merged already (just add the eex test probably) :)

Otherwise when I get some time I can continue tinkering with it

**Comment by josevalim** [comment]
Sounds good. I will have some eyes on this too then!

**Comment by josevalim** [comment]
@viniciusmuller the tests are failing, can you please take a look? :)

**Comment by josevalim** [comment]
Nevermind, it is failing exactly because there is something unclear. I will investigate. :)

**Comment by josevalim** [comment]
Manually merged with a fix! :green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request closed without merging**
